pkg/agent/agentclient/stream.go

```diff
@@ -168,48 +207,164 @@ func (c *RedialableAgentClient) Recv() (*agent.Packet, error) {
 	return pkt, nil
 }
 
-func (c *RedialableAgentClient) Connect() error {
-	if err := c.tryConnect(); err != nil {
-		return err
+// Connect makes the grpc dial to the proxy server. It returns the serverID
+// it connects to.
+func (c *RedialableAgentClient) Connect() (string, error) {
+	var err error
+	r, err := c.tryConnect()
+	if err != nil {
+		return "", err
 	}
+	c.serverID = r.serverID
+	c.serverCount = r.serverCount
+	c.conn = r.grpcConn
+	c.stream = r.agentServiceClient
+	return c.serverID, nil
+}
 
-	go c.probe()
-	return nil
+// The goal is to make the chance that client's Connect rpc call has never hit
+// the wanted server after "retries" times to be lower than 10^-2.
+func retryLimit(serverCount int) (retries int) {
+	switch serverCount {
+	case 1:
+		return 3 // to overcome transient errors
+	case 2:
+		return 3 + 7
+	case 3:
+		return 3 + 12
+	case 4:
+		return 3 + 17
+	case 5:
+		return 3 + 21
+	default:
+		// we don't expect HA server with more than 5 instances.
+		return 3 + 21
+	}
 }
 
 func (c *RedialableAgentClient) reconnect() {
-	klog.Info("start to connect...")
-
-	var err error
-	var retry int
-
-	for retry < c.Retry {
-		if err = c.tryConnect(); err == nil {
-			klog.Info("connected")
+	klog.Info("start to reconnect...")
+
+	var retry, limit int
+
+	limit = retryLimit(c.serverCount)
+	for retry < limit {
+		r, err := c.tryConnect()
+		if err != nil {
+			retry++
+			klog.Infof("Failed to connect to proxy server, retry %d in %v: %v", retry, c.Interval, err)
+			time.Sleep(c.Interval)
+			continue
+		}
+		switch {
+		case r.serverID == c.serverID:
+			klog.Infof("reconnected to %s", r.serverID)
+			c.conn = r.grpcConn
+			c.stream = r.agentServiceClient
 			c.doneReconnect(nil)
 			return
+
+		case r.serverID != c.serverID && c.cs.HasID(r.serverID):
+			// reset the connection
+			err := r.grpcConn.Close()
+			if err != nil {
+				klog.Infof("failed to close connection to %s: %v", r.serverID, err)
+			}
+			retry++
+			klog.Infof("Trying to reconnect to proxy server %s, got connected to proxy server %s, for which there is already a connection, retry %d in %v", c.serverID, r.serverID, retry, c.Interval)
+			time.Sleep(c.Interval)
+		case r.serverID != c.serverID && !c.cs.HasID(r.serverID):
+			// create a new client
+			cc := copyRedialableAgentClient(*c)
+			cc.stream = r.agentServiceClient
+			cc.conn = r.grpcConn
+			ac := &AgentClient{
+				connContext: make(map[int64]*connContext),
+				stream:      &cc,
+				serverID:    cc.serverID,
```

Comment:
This should be `r.serverID`, else the new client would still try to reconnect to the old `cc.serverID` when attempting another reconnect.